### ISSUE TITLE
MQTT Egress and Ingress to same topic structure causes looping

### ISSUE THREAD

**Issue by hellkrusher** [emqx/emqx#13511]
### What happened?

When publishing to the bridge broker or local broker an infinite publish loop occurs when using a shared topic. 

### What did you expect to happen?

I should be able to publish to any topic and have it republish to the remote broker only once and vice versa. I'm trying to set up a Unified Namespace so I need shared topic structures amongst the different brokers in the ISA-95 part 2 hierarchy. 

### How can we reproduce it (as minimally and precisely as possible)?

I created an MQTT broker connector to an external emqx broker with the `Bridge Mode` enabled.
I created an ingress flow from `MQTT Broker` to `Republish`
I created an egress flow from `Messages` to `MQTT Broker`
I publish a message to a topic and an infinite publish loop occurs.
If I disabled the egress the ingress works as expected
If I disable the ingress the ingress works as expected. 
This Bridging with topic structure sharing was working in 5.4 when there was a separate data bridge section in the GUI. 

### EMQX version

5.7.1 (Open Source)

### OS version

RHEL 8.10

### Log files

<details>

</details>

**Comment by qzhuyan** [comment]
Looks like you created a loopback indeed.  
Both brokers are EMQX? same version? I remember we have some magic bits to avoid loop.
could you provide the configuration for me to understand?

**Comment by zmstone** [comment]
It did not loop in 5.4 probably because the message happed to be published and subscribed by one client, so the "bridge mode" or "no local" (mqtt v5) flag took effect.

The problem is the remote broker is unable to tell if the publisher and the subscriber are the same entity because they could be:
- two clients on the same node
- two clients from different nodes in the same cluster

There is little we can do when the remote is not EMQX.
For EMQX as remote broker, we'll try to make use of some message properties etc to break the loop.
I am also experimenting with a workaround for this.

**Comment by zmstone** [comment]
I reproduced it. @qzhuyan

**Comment by hellkrusher** [comment]
> Looks like you created a loopback indeed. Both brokers are EMQX? same version? I remember we have some magic bits to avoid loop. could you provide the configuration for me to understand?

Yes exact same version and setup just on a different server. If there is any setup or workaround to implement bridging on a shared topic please let me know. I realize that with the current setup an infinite loop is what should happen but was hoping there was some mechanism to prevent this like in previous versions.

**Comment by zmstone** [comment]
The workaround is: create two rules:

1. forward messages to peer EMQX cluster with a `forward/` prefix mount.
2. unmount `forward/` prefix for messages matching `forward/#`.

In order for the workaround to work, I need to make a change in the code.
https://github.com/emqx/emqx/pull/13516
An example config is provided in the PR.

**Comment by qzhuyan** [comment]
Just some idea:  have some 'max-hop-count' private header in Publish message and set it to one to avoid the message loop back. 

This could be added by client or by the bridge client in the broker.

**Comment by zmstone** [comment]
> Just some idea: have some 'max-hop-count' private header in Publish message and set it to one to avoid the message loop back.
> 
> This could be added by client or by the bridge client in the broker.

That's can be useful to break loop when bridging towards other vendors (given that they preserve the header). 
This feature is to break a infinite recursion of hook callback.

**Comment by zmstone** [comment]
> Just some idea: have some 'max-hop-count' private header in Publish message and set it to one to avoid the message loop back.
> 
> This could be added by client or by the bridge client in the broker.

It can be useful towards other vendors (given that they preserve the header).
However it still would waste one hop of delivery (only the message is looped back to EMQX can we know should be discarded).

**Comment by hellkrusher** [comment]
I've given the recommended setup a try suggested in PR #13516 . It works as described but unfortunately we would need  `${qos}` and `${flags.retain}` passed thru or else this wouldn't be of much use in manufacturing (such as publishing static Asset IDs once, and having those cascaded down to the line level broker topics for future subscriber use). Also, we would want to bridge each security zone so going from `Enterprise -> Site -> Area -> Line` , so the topics would need to take multiple hops, not just republish to a single broker, then stop. We wouldn't want to publish from Line level broker to Enterprise level broker directly for security and architectural reasons. 

Is forming a cluster with all brokers an option to achieving a UNS Architecture? My preference would be to have bridging functionality working again but if that's not possible just wondering if clustering would be the only other way and if there are performance considerations for doing this for all brokers in the enterprise.

**Comment by zmstone** [comment]
Hi @hellkrusher 

You should be able to pass through QoS without any problem today.
To preserve `retain` flag is not possible at the moment but not complex to fix --- however no eta can be provided now.

If your security zone are distributed in different regions geographically, the regular clustering is not recommended.
Enterprise core-replica clustering will work:
https://docs.emqx.com/en/emqx/latest/design/clustering.html#core-and-replicant-nodes
There is a blog for ref:
https://www.emqx.com/en/blog/exploring-geo-distribution-in-emqx-for-enhanced-scalability

**Comment by hellkrusher** [comment]
Thank you for the clarification. One thing I noticed is, in a cluster, if a client is subscribed to a topic on Node A and publishes to that same topic on Node C but if Node A losses it's connection to Node C the message is not delivered. Is there any feature that would route the message from Node C to Node B, then to Node A or is a cluster always point to point? I noticed the retained message is available if Node A and Node C are not connected. Is this where a linked cluster would come in to bridge each security zone?  
In summary my question is what should happened when a publish and subscribe client are on two different nodes and they loose their direct connection to each other?

**Comment by zmstone** [comment]
Hi @hellkrusher 

In case of clustering, when two nodes cannot connect, then the message will be lost because between the nodes, it's a pool of TCP connections for message forwarding. (The protocol is a simple packet-4 framing with Erlang's term serialization, but not MQTT messages).

Clustering linking makes use of a pool of MQTT connections to forward data on MQTT protocol itself.
Each connection has a buffer to support short-time connection loss between the linked nodes, it should send buffered messages after connection is re-established.
The buffer is in memory though, so it cannot hold data for long, messages will be dropped after buffer overflow.

**Comment by zmstone** [comment]
closing it because now EMQX has:

- Loop breaker in republish action: https://github.com/emqx/emqx/pull/13516, so EMQX egress bridges can publish to each other without looping.
- RAP support, so ingest/subscribe bridge can republish with retained: https://github.com/emqx/emqx/pull/17481
- [Cluster Linking Feature](https://docs.emqx.com/en/cloud/latest/cluster_linking/cluster_linking.html), so EMQX clusters can share local subscriptions with each other, then only relevant messages are replicated.
